snuba/datasets/entities/transactions.py

```diff
@@ -86,20 +98,89 @@
 )
 
 
+class SampledSimplePipelineBuilder(SimplePipelineBuilder):
+    def build_planner(
+        self, query: LogicalQuery, settings: RequestSettings,
+    ) -> EntityQueryPlanner:
+        new_query = deepcopy(query)
+        new_query.set_sample(0.1)
+        return super().build_planner(new_query, settings)
+
+
 class BaseTransactionsEntity(Entity, ABC):
     def __init__(self, custom_mappers: Optional[TranslationMappers] = None) -> None:
         storage = get_writable_storage(StorageKey.TRANSACTIONS)
         schema = storage.get_table_writer().get_schema()
 
+        pipeline_builder = SimplePipelineBuilder(
+            query_plan_builder=SingleStorageQueryPlanBuilder(
+                storage=storage,
+                mappers=transaction_translator
+                if custom_mappers is None
+                else transaction_translator.concat(custom_mappers),
+            ),
+        )
+        sampled_pipeline_builder = SampledSimplePipelineBuilder(
+            query_plan_builder=SingleStorageQueryPlanBuilder(
+                storage=storage,
+                mappers=transaction_translator
+                if custom_mappers is None
+                else transaction_translator.concat(custom_mappers),
+            ),
+        )
+
+        def selector_func(query: LogicalQuery, referrer: str) -> Tuple[str, List[str]]:
+            if referrer == "tagstore.__get_tag_keys":
+                sample_query_rate = state.get_config("snuplicator_sampled_query", 0.0)
+                assert isinstance(sample_query_rate, float)
+                if random.random() < sample_query_rate:
+                    return "transactions", ["sampler"]
+
+            return "transactions", []
+
+        def callback_func(
+            query: LogicalQuery, referrer: str, results: List[Result[QueryResult]],
+        ) -> None:
+            if not results:
+                metrics.increment(
+                    "query_result",
+                    tags={"match": "empty", "primary": "none", "referrer": referrer},
+                )
+                return
+
+            primary_result = results.pop(0)
+            primary_function_id = primary_result.function_id
+            secondary_result = results.pop(0) if len(results) > 1 else None
+
+            # compare results
+            comparison = "one_result"
+            if secondary_result:
+                primary_result_data = primary_result.result.result["data"]
+                secondary_result_data = secondary_result.result.result["data"]
+                if secondary_result_data == primary_result_data:
```

Comment:
I suspect you'll hit the same issue I did with nan != nan

Reply:
This is never going to match without adjusting the results on the sampled query.
Passing a sampling rate to Clickhouse will simply skip data blocks. Clickhouse is not going to to multiply the results by 1/sample_rate for you.

```
SELECT count()
FROM transactions_local
WHERE (project_id = SOMETHING) AND (finish_ts > (now() - (24 * 3600)))

┌──count()─┐
│ 14291614 │
└──────────┘
```

```
SELECT count()
FROM transactions_dist
SAMPLE 1 / 10
WHERE (project_id = SOMETHING) AND (finish_ts > (now() - (24 * 3600)))

┌─count()─┐
│ 1429359 │
└─────────┘
```

Also even if you adjust the results, you are unlikely to ever hit exactly the same number, I think it would be better to call it a match if the adjusted sampled result is within a percentage of the original one (like 1-5%).

Reply:
Ah I forgot there was a count in this query. Yeah OK I'll rethink this.